profiles/views.py

```diff
@@ -37,6 +46,96 @@ class ProfileViewSet(RetrieveModelMixin, UpdateModelMixin, GenericViewSet):
+
+        Args:
+            url (str): mailgun api url:
+            email (str): user email
+        """
+        try:
+            response = requests.post(
+                url,
+                auth=('api', settings.MAILGUN_KEY),
+                data={
+                    'address': email,
+                    'tag': '*'
+                }
+            )
+        except requests.exceptions.RequestException:
+            log.exception(
+                "Unable to add email: %s to mailgun unsubscribes list.", email
+            )
+            return
+        if response.status_code == statuses.HTTP_200_OK:
+            log.debug(
+                "Added user's email: %s to mailgun unsubscribes list. Message received: %s",
+                email,
+                response.json()
+            )
+
+    @staticmethod
+    def remove_email_from_unsub_list(url, email):
+        """
+        It removes user email from mailgun unsub list.
+
+        Args:
+            url (str): mailgun api url:
+            email (str): user email
+        """
+        try:
+            response = requests.delete(url, auth=('api', settings.MAILGUN_KEY))
+        except requests.exceptions.RequestException:
+            log.exception(
+                "Unable to remove email: %s from mailgun unsubscribes list.", email
+            )
+            return
+        if response.status_code == statuses.HTTP_200_OK:
+            log.debug(
+                "Removed user's email: %s from mailgun unsubscribes list. Message received: %s",
+                email,
+                response.json()
+            )
+
+    @staticmethod
+    def mailgun_action(email_optin, email):
+        """
+        it removes user from mailgun unsubscribes list.
+        https://documentation.mailgun.com/en/latest/api-suppressions.html#delete-a-single-unsubscribe
+
+        Args:
+            email_optin (bool): email optin flag
+            email (str): user email
+        """
+        url = "{base}/unsubscribes".format(base=settings.MAILGUN_URL)
+        if email_optin:
+            url = '{base}/{email}'.format(base=url, email=email)
+            ProfileViewSet.remove_email_from_unsub_list(url, email)
+        else:
+            ProfileViewSet.add_email_to_unsub_list(url, email)
+
+    @staticmethod
+    def simple_response(status):
+        """
+        returns status response.
+
+        Args:
+            status (int): Http status
+        """
+        return HttpResponse(status=status)
+
+    def update(self, request, *args, **kwargs):
+        """
+        updates user profile.
+        """
+        if 'email_optin' in request.data and 'email' in request.data:
+            # perform mailgun action if email optin is set then remove user from mailgun unsubscription
+            # list otherwise add user to mailgun unsubscription.
+            ProfileViewSet.mailgun_action(request.data['email_optin'], request.data['email'])
+        return super().update(request, *args, **kwargs)
```

Comment:
I think we can override DRF's `perform_update` method (documentation in [this section](http://www.django-rest-framework.org/api-guide/generic-views/#methods)) instead of overriding `update`. `perform_update` exists specifically to override and provide some custom saving functionality. Might look something like this:

```python
def perform_update(self, serializer):
  profile = serializer.save()
  if 'email_optin' in serializer.validated_data:
    MailgunClient.toggle_unsubscription(profile, serializer.validated_data['email_optin'])
```